Typescript saga immutablejs

bflamand opened the pull request:
<!-- Enter details of the change here. Include additional tests that have been done, reference to the issue for tracking, etc. -->

<!-- Please check the completed items below -->
### PR checklist
 
- [ ] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ ] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [ ] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) beforehand.
- [ ] Run the shell script `./bin/generate-samples.sh`to update all Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. These must match the expectations made by your contribution. You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ ] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`
- [ ] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

bflamand-work committed:
first commit: add cli option for saga and records. Added dummy sagas.mustache test file.

bflamand-work committed:
More progress with default values. First prototype for isEntity and isUniqueId.

bflamand-work committed:
record generation complete

bflamand-work committed:
record generation complete

bflamand-work committed:
progress with saga generation

bflamand-work committed:
progress with saga generation

bflamand-work committed:
first fully working saga generation

bflamand-work committed:
Merge remote-tracking branch 'origin/master' into typescript-saga-immutablejs

# Conflicts:
#	modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenModel.java
#	modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenParameter.java
#	modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java
#	modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptFetchClientCodegen.java
#	modules/openapi-generator/src/main/resources/typescript-fetch/models.index.mustache
#	modules/openapi-generator/src/test/java/org/openapitools/codegen/options/TypeScriptFetchClientOptionsProvider.java
#	modules/openapi-generator/src/test/java/org/openapitools/codegen/typescript/fetch/TypeScriptFetchClientOptionsTest.java

bflamand-work committed:
merge with latest master

JFCote commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java:
```diff
@@ -38,6 +38,11 @@
+     * Alternative version of the language-specific data type when generator needs a different output in some templates.
+     * For example this is used by typescript fetch generator to output either Array (javascript) or List (immutablejs)
+     */
+    public String dataTypeAlternate;
```
You should not do this. This is a common thing to do in most of the existing template. You need to deal with this in the TypescriptFetchCodegen.java. You can you typeMapping or even postProcessing. But don't create yet another property

bflamand replied:
From what I understand, the only way to have another property accessible in the mustache file for models is to add it here? If there was another way, this would be great, we would not have to "polute" global codegen here...
Btw, this would not be needed if we made a new generator specifically for outputing sagas&records. But here we need to continue to output the normal typescript fetch in addition to the new stuff so we need a second and different dataType here for the new stuff. This would also not be needed if mustache fully supported if/else but it does not. So we need to do the if/else in code and prepare this variable so it can be used in records and saga mustache files.

JFCote commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java:
```diff
@@ -108,6 +113,15 @@
+    /**
+     * The property represents a unique id (x-isUniqueId: true); field populated only by typescript fetch for now.
+     */
+    public boolean isUniqueId;
```
is there a common way to name  properties that are on the vendor side so people are not tempted to use them in their template by error?

bflamand replied:
I looked and did not find a pattern for this. Naming seams to be different depending on who added a variable. But we could at least use one for our code and remain consistant.

JFCote commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenProperty.java:
```diff
@@ -158,6 +172,11 @@
     // the undeclared properties. 
     public CodegenProperty items;
+
+    public String itemsDataType; // if isContainer == true, this is the dataType of the items else it is null. Used by typescript fetch only for now.
```
Don't use these 3 tags since they have duplicated information.
itemsDataType: This is already in the item, you don't need this.
itemsAreModels: You can do "if" in your template based on the isUniqueId
itemsAreEntities: You can do "if" in your template based on the item type.  This is very common in most template.

bflamand replied:
again, this was added because I did not find any way to get this information once in the mustache file. I needed to expose it here.

JFCote commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java:
```diff
@@ -5450,12 +5450,7 @@ public void updateCodegenPropertyEnum(CodegenProperty var) {
-            } else {
-                enumDefaultValue = var.defaultValue;
-            }
+			final String enumDefaultValue = var.defaultValue;
```
Why did you remove this?

bflamand replied:
I found a bug!

JFCote commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractTypeScriptClientCodegen.java:
```diff
@@ -621,11 +621,6 @@ public String toEnumValue(String value, String datatype) {
         }
     }
 
-    @Override
```
same question here, why did you remove this?

JFCote commented on modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache:
```diff
@@ -0,0 +1,148 @@
+	recType: "{{classname}}Record" as "{{classname}}Record",
+{{#vars}}
+    {{#isArray}}
+    {{#itemsAreModels}}
```
This is directly related to what I have said above, you don't need these 3 new properties.
The logic is: 
* All items are models unless they are native type
* You can know that a model is an "entity" if isUnique is true.

bflamand replied:
Will check if {{^nativeType}} is the same as itemsAreModels ...

JFCote commented on modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache:
```diff
@@ -0,0 +1,148 @@
+	public normalize(apiObject: {{classname}}, asEntity?: boolean): {{classname}} {
+		(apiObject as any).recType = {{#isEntity}}asEntity ? "{{classname}}RecordEntity" : {{/isEntity}}"{{classname}}Record";
+{{#vars}}
+        {{#isUniqueId}}
```
you are already doing it the right way here

JFCote commented on modules/openapi-generator/src/main/resources/typescript-fetch/recordGeneric.mustache:
```diff
@@ -0,0 +1,148 @@
+{{#isMetaDataResponse}}
+{{#vars.1}}
+
+	public fromApi_data(apiObject: {{classname}}): {{{dataTypeAlternate}}} {
```
like I said above, all this thing with the dataTypeAlternate must be  handled in the codegen code

JFCote commented on modules/openapi-generator/src/main/resources/typescript-fetch/runtimeSagasAndRecords.mustache:
```diff
@@ -0,0 +1,115 @@
+export const knownIndexedSetByKey = [];
+
+export function appFromJS(any: any): any {
+	return originalFromJS(any, (key, value) => {
```
please review all your file and make sure you use a "4 space tabs" instead of "real tabs". this is the standard in this project

bflamand replied:
oh...

JFCote commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenOperation.java:
```diff
@@ -29,9 +29,9 @@
-            isRestful, isDeprecated, isCallbackRequest, uniqueItems;
-    public String path, operationId, returnType, returnFormat, httpMethod, returnBaseType,
-            returnContainer, summary, unescapedNotes, notes, baseName, defaultResponse; 
+            isRestful, isDeprecated, isCallbackRequest, uniqueItems, returnTypeIsMetaDataResponse, returnTypeIsMetaOnlyResponse, returnTypeSupportsEntities, returnTypeIsModel, returnTypeIsArray;
```
it kind of weird to have a concept of entity in a very generic object called CodegenOperation. I think this should be handled in the specific codegen in the java code. I don't think any of these variable should be here.

JFCote replied:
Maybe follow the pattern here?
`class ExtendedCodegenOperation extends CodegenOperation {`

JFCote reviewed (changes requested):
Lot of things to remove and/or move

bflamand-work committed:
removed unneeded "items" properties.

bflamand-work committed:
moved global CodegenModel modifications into subclass ExtendedCodegenModel used exclusively by TypescriptFetchClient. Adding missing samples files.

bflamand-work committed:
moved global CodegenOperation modifications into subclass ExtendedCodegenOperation used exclusively by TypescriptFetchClient.

bflamand-work committed:
moved global CodegenProperty modifications into subclass ExtendedCodegenProperty used exclusively by TypescriptFetchClient.

bflamand-work committed:
moved global CodegenParameter modifications into subclass ExtendedCodegenParameter used exclusively by TypescriptFetchClient.

bflamand-work committed:
added the missing "allSagas" export.

bflamand-work committed:
renamed & reworked "meta data response" flags to a more useful general concept of "operation return passthrough"

bflamand-work committed:
added vendor flag keepAsJSObject as escape hatch to support circular dependencies in models and other special cases. Also fixed issues with default values for some records properties.

bflamand-work committed:
added autodetection for passthrough to simplify standardised specs.

bflamand-work committed:
fix small issue with passthrough void

bflamand-work committed:
fix small issues with passthrough void and missing passthrough imports in some cases. Fix issues with enum default values.

bflamand-work committed:
fix small issues with passthrough void and missing passthrough imports in some cases. Fix issues with enum default values.

bflamand-work committed:
Merge remote-tracking branch 'origin/master' into typescript-saga-immutablejs

bflamand-work committed:
Added "reservedRecordField" feature to support remapping fields names that cannot be used in Records. Added missing export to record: toApi().

bflamand-work committed:
added uniqueId inference. Fix small generation when uniqueId property is an array.

bflamand-work committed:
removed feature "reservedRecordField" and replaced it with existing built-in "reserved words" feature. Fix minor issues with typings in generated files.

bflamand-work committed:
Changed api recType names to make them less likely to cause name conflicts. Added generated ApiEntities (record, reducer & selector) files.

bflamand-work committed:
Moved location of ApiEntities related files and fix issues with exports.

bflamand-work committed:
merge master

bflamand-work committed:
- merge latest master
- renamed fake test apis to better fit the "pet theme"
- added mode for "SourceOnlyLibrary" (same as used in codegen typescript jquery)

bflamand-work committed:
- missing ganarate sampless

bflamand-work committed:
Merge remote-tracking branch 'origin/master' into typescript-saga-immutablejs

bflamand-work committed:
- Modified way to export apiEntitiesSelectpr to reduce typescript analysis time for consuming project. Removed tab characters in mustache files. Reformat code for TypeScriptFetchClientCodegen to try to remove false positive for tabs vs spaces.

bflamand-work committed:
- added markErrorsAsHandled property to api sagas. Increased typescript version to address some typing errors on library build.

bflamand-work committed:
- fix bug in saga interfaces. Upgraded to typescript "strict" mode to ensure proper typechecking info is generated.

bflamand-work committed:
- added optional id for apiEntity selectors. Added toInlined() support to convert an entity to an inlined model recursively.

bflamand-work committed:
- minor tweak for apiEntitySelector to accept null id

bflamand-work committed:
- minor tweak for apiEntitySelector

bflamand-work committed:
Merge remote-tracking branch 'origin/master' into typescript-saga-immutablejs

# Conflicts:
#	modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/TypeScriptFetchClientCodegen.java

bflamand-work committed:
- runned ensure up to date.

bflamand-work committed:
Revert "- runned ensure up to date."

This reverts commit ea9b4aed

bflamand-work committed:
- runned ensure up to date.

bflamand-work committed:
- runned ensure up to date.
